link do dataset: https://www.kaggle.com/laotse/credit-risk-dataset?select=credit_risk_dataset.csv

## Descrição das colunas :


person_age: idade da pesoa quando pegou o empréstimo.

person_income: renda por ano.

person_home_ownership: tipo de propriedade da casa.

person_emp_length: quantidade de tempo em anos que a pessoa está empregada.

loan_intent: objetivo do emprestimo.

loan_grade: é um sistema de classificação que envolve a atribuição de uma pontuação de qualidade a um empréstimo com base no histórico de crédito do mutuário, qualidade da caução e probabilidade de reembolso do principal e juros.

loan_amnt:valor do empréstimo.

loan_int_rate: o juros pago pelo empréstimo.

loan_status: status de risco do empréstimo.

loan_percent_income: relação entre o valor do empréstimo e a receita anual.

cb_person_default_on_file: histórico de inadimplência.

cb_person_cred_hist_length: representa o número de anos de história pessoal desde o primeiro empréstimo tomado por uma pessoa.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('credit_risk_dataset.csv')

In [3]:
df.head()

person_age  person_income person_home_ownership  person_emp_length  \
0          22          59000                  RENT              123.0   
1          21           9600                   OWN                5.0   
2          25           9600              MORTGAGE                1.0   
3          23          65500                  RENT                4.0   
4          24          54400                  RENT                8.0   

  loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
0    PERSONAL          D      35000          16.02            1   
1   EDUCATION          B       1000          11.14            0   
2     MEDICAL          C       5500          12.87            1   
3     MEDICAL          C      35000          15.23            1   
4     MEDICAL          C      35000          14.27            1   

   loan_percent_income cb_person_default_on_file  cb_person_cred_hist_length  
0                 0.59                         Y                           3  
1                 0.10                         N                           2  
2                 0.57                         N                           3  
3                 0.53                         N                           2  
4                 0.55                         Y                           4

In [4]:
df.columns = [ 'idade','renda_por_ano','tipo_moradia','tempo_de_emprego','objetivo_emprestimo',
               'class_risco_historico','montante_emprestimo','taxa_juros','risco_emprestimo',
               'porc_empr_renda_anual','inadimplencia_hist','qtd_anos_primeiro_emprestimo']

In [5]:
df = df[['idade','tempo_de_emprego','tipo_moradia',
         'objetivo_emprestimo','montante_emprestimo','taxa_juros',
         'renda_por_ano','porc_empr_renda_anual','class_risco_historico','inadimplencia_hist',
         'qtd_anos_primeiro_emprestimo','risco_emprestimo']]

In [6]:
df.head()

idade  tempo_de_emprego tipo_moradia objetivo_emprestimo  \
0     22             123.0         RENT            PERSONAL   
1     21               5.0          OWN           EDUCATION   
2     25               1.0     MORTGAGE             MEDICAL   
3     23               4.0         RENT             MEDICAL   
4     24               8.0         RENT             MEDICAL   

   montante_emprestimo  taxa_juros  renda_por_ano  porc_empr_renda_anual  \
0                35000       16.02          59000                   0.59   
1                 1000       11.14           9600                   0.10   
2                 5500       12.87           9600                   0.57   
3                35000       15.23          65500                   0.53   
4                35000       14.27          54400                   0.55   

  class_risco_historico inadimplencia_hist  qtd_anos_primeiro_emprestimo  \
0                     D                  Y                             3   
1                     B                  N                             2   
2                     C                  N                             3   
3                     C                  N                             2   
4                     C                  Y                             4   

   risco_emprestimo  
0                 1  
1                 0  
2                 1  
3                 1  
4                 1

In [7]:
df.shape

(32581, 12)

# Tratando os dados

- Tratando os dados ausentes.

In [8]:
df.isnull().sum()

idade                              0
tempo_de_emprego                 895
tipo_moradia                       0
objetivo_emprestimo                0
montante_emprestimo                0
taxa_juros                      3116
renda_por_ano                      0
porc_empr_renda_anual              0
class_risco_historico              0
inadimplencia_hist                 0
qtd_anos_primeiro_emprestimo       0
risco_emprestimo                   0
dtype: int64

In [9]:
df['taxa_juros'] = df['taxa_juros'].fillna(df.taxa_juros.mean())

In [10]:
df = df.dropna()

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31686 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   idade                         31686 non-null  int64  
 1   tempo_de_emprego              31686 non-null  float64
 2   tipo_moradia                  31686 non-null  object 
 3   objetivo_emprestimo           31686 non-null  object 
 4   montante_emprestimo           31686 non-null  int64  
 5   taxa_juros                    31686 non-null  float64
 6   renda_por_ano                 31686 non-null  int64  
 7   porc_empr_renda_anual         31686 non-null  float64
 8   class_risco_historico         31686 non-null  object 
 9   inadimplencia_hist            31686 non-null  object 
 10  qtd_anos_primeiro_emprestimo  31686 non-null  int64  
 11  risco_emprestimo              31686 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.1+ MB


- Tratando os dados das colunas **objetivo_emprestimo**,**tipo_moradia** e **inadimplencia_hist**

In [12]:
objetivo = {'EDUCATION': "educacao", 'MEDICAL': "saude", 'VENTURE': "risco", 
            'PERSONAL': "pessoal", 'DEBTCONSOLIDATION': "consolidacao_debito" ,'HOMEIMPROVEMENT':'reformar_casa'}
df['objetivo_emprestimo'] = df['objetivo_emprestimo'].map(objetivo)


tipo_moradia = {'RENT': 'alugada', 'MORTGAGE': 'hipoteca', 'OWN': 'propria', 'OTHER': 'outro'  }
df['tipo_moradia'] = df['tipo_moradia'].map(tipo_moradia)

In [13]:
df['inadimplencia_hist'] = df['inadimplencia_hist'].apply(lambda x:  1 if x =='Y' else 0)

In [14]:
df.head()

idade  tempo_de_emprego tipo_moradia objetivo_emprestimo  \
0     22             123.0      alugada             pessoal   
1     21               5.0      propria            educacao   
2     25               1.0     hipoteca               saude   
3     23               4.0      alugada               saude   
4     24               8.0      alugada               saude   

   montante_emprestimo  taxa_juros  renda_por_ano  porc_empr_renda_anual  \
0                35000       16.02          59000                   0.59   
1                 1000       11.14           9600                   0.10   
2                 5500       12.87           9600                   0.57   
3                35000       15.23          65500                   0.53   
4                35000       14.27          54400                   0.55   

  class_risco_historico  inadimplencia_hist  qtd_anos_primeiro_emprestimo  \
0                     D                   1                             3   
1                     B                   0                             2   
2                     C                   0                             3   
3                     C                   0                             2   
4                     C                   1                             4   

   risco_emprestimo  
0                 1  
1                 0  
2                 1  
3                 1  
4                 1

# EDA

- Quais fatores influenciam na classificação do empréstimo como sendo de risco ou não.




- Transformar variáveis categóricas em variáveis numéricas.

In [15]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()


df['tipo_moradia_id']  = encoder.fit_transform(df['tipo_moradia'])
df['objetivo_emprestimo_id']  = encoder.fit_transform(df['objetivo_emprestimo'])
df['class_risco_historico_id']  = encoder.fit_transform(df['class_risco_historico'])

- Verificando dados inconsistentes.

In [16]:
df.describe().T

count          mean           std      min  \
idade                         31686.0     27.747302      6.360283    20.00   
tempo_de_emprego              31686.0      4.789686      4.142630     0.00   
montante_emprestimo           31686.0   9661.337815   6336.226458   500.00   
taxa_juros                    31686.0     11.037157      3.070128     5.42   
renda_por_ano                 31686.0  66691.878306  62427.286689  4000.00   
porc_empr_renda_anual         31686.0      0.169612      0.106297     0.00   
inadimplencia_hist            31686.0      0.177649      0.382223     0.00   
qtd_anos_primeiro_emprestimo  31686.0      5.809316      4.060867     2.00   
risco_emprestimo              31686.0      0.215426      0.411124     0.00   
tipo_moradia_id               31686.0      0.648141      0.831374     0.00   
objetivo_emprestimo_id        31686.0      2.501641      1.753100     0.00   
class_risco_historico_id      31686.0      1.226662      1.168090     0.00   

                                   25%           50%       75%         max  
idade                            23.00     26.000000     30.00      144.00  
tempo_de_emprego                  2.00      4.000000      7.00      123.00  
montante_emprestimo            5000.00   8000.000000  12500.00    35000.00  
taxa_juros                        8.49     11.011695     13.16       23.22  
renda_por_ano                 39397.00  56000.000000  80000.00  6000000.00  
porc_empr_renda_anual             0.09      0.150000      0.23        0.83  
inadimplencia_hist                0.00      0.000000      0.00        1.00  
qtd_anos_primeiro_emprestimo      3.00      4.000000      8.00       30.00  
risco_emprestimo                  0.00      0.000000      0.00        1.00  
tipo_moradia_id                   0.00      0.000000      1.00        3.00  
objetivo_emprestimo_id            1.00      2.000000      4.00        5.00  
class_risco_historico_id          0.00      1.000000      2.00        6.00

- Pode ser observados dados inconsistentes, como por exemplo idade máxima de 144 anos e tempo de emprego de 123 anos

- Tratando os dados inconsistentes das colunas **idade** e **tempo_de_emprego**.

In [17]:
df.query('idade > 94')

idade  tempo_de_emprego tipo_moradia objetivo_emprestimo  \
81       144               4.0      alugada               risco   
183      144               4.0     hipoteca            educacao   
575      123               2.0      alugada            educacao   
747      123               7.0      alugada               risco   
32297    144              12.0     hipoteca             pessoal   

       montante_emprestimo  taxa_juros  renda_por_ano  porc_empr_renda_anual  \
81                    4800   13.570000         250000                   0.02   
183                   6000   11.860000         200000                   0.03   
575                  20400   10.250000          80004                   0.25   
747                  20000   11.011695          78000                   0.26   
32297                 5000   12.730000        6000000                   0.00   

      class_risco_historico  inadimplencia_hist  qtd_anos_primeiro_emprestimo  \
81                        C                   0                             3   
183                       B                   0                             2   
575                       B                   0                             3   
747                       B                   0                             4   
32297                     C                   0                            25   

       risco_emprestimo  tipo_moradia_id  objetivo_emprestimo_id  \
81                    0                0                       4   
183                   0                1                       1   
575                   0                0                       1   
747                   0                0                       4   
32297                 0                1                       2   

       class_risco_historico_id  
81                            2  
183                           1  
575                           1  
747                           1  
32297                         2

In [20]:
df.query('tempo_de_emprego >40')

idade  tempo_de_emprego tipo_moradia objetivo_emprestimo  \
0         22             123.0      alugada             pessoal   
210       21             123.0     hipoteca               risco   
32355     78              41.0      alugada               saude   

       montante_emprestimo  taxa_juros  renda_por_ano  porc_empr_renda_anual  \
0                    35000       16.02          59000                   0.59   
210                  20000        6.54         192000                   0.10   
32355                 3000        7.51          48000                   0.06   

      class_risco_historico  inadimplencia_hist  qtd_anos_primeiro_emprestimo  \
0                         D                   1                             3   
210                       A                   0                             4   
32355                     A                   0                            25   

       risco_emprestimo  tipo_moradia_id  objetivo_emprestimo_id  \
0                     1                0                       2   
210                   0                1                       4   
32355                 0                0                       5   

       class_risco_historico_id  
0                             3  
210                           0  
32355                         0

In [21]:
df = df.query('idade <= 94 and tempo_de_emprego <= 41')

In [75]:
df.to_csv('dados_analise_risco_de_cresdito.csv',index=False)